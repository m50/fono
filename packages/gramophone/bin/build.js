#!/usr/bin/env node
/* eslint-disable import/no-extraneous-dependencies */
const babel = require('@babel/core');
const glob = require('glob-promise');
const { mkdir, readFile, rmdir, writeFile } = require('fs/promises');
const { existsSync } = require('fs');
const { dirname } = require('path');
const babelConfig = require('../config/.babelrc.json');
const esbuildConfig = require('../config/.esbuild.json');
const tsconfigRaw = require('../tsconfig.json');
const esbuild = require('esbuild');
const { relative, resolve } = require('path');

const { argv } = process;
if (argv[argv.length - 1] === '') {
  argv.pop();
}
const arg = argv.pop();
if (arg === '--help') {
  console.log('\n\tJust run build on it\'s own. No additional arguments available.');
  console.log('\tAlternatively, you can use `build:prod` to conduct a production build.\n');
  process.exit(0);
}
if (!process.env.NODE_ENV) {
  process.env.NODE_ENV = 'development';
}
if (arg === 'prod') {
  process.env.NODE_ENV = 'production';
}

const rootDir = dirname(__dirname);
const srcDir = `${rootDir}/src`;
const distDir = `${rootDir}/dist`;

async function transformEsbuild(source, path, paths) {
  const output = await esbuild.transform(source, {
    ...esbuildConfig,
    sourcefile: path,
    minify: process.env.NODE_ENV === 'production',
    tsconfigRaw,
  });
  const s = process.env.NODE_ENV === 'production' ? '' : ' '
  if (output.warnings.length > 0) {
    throw output.warnings;
  }
  output.code = `"use strict";${output.code}`
    .replace(/import (.*?) from\s*(['"].*?['"]);?/g, `const $1${s}=${s}require($2);`)
    .replace(/import\s*(['"].*?['"]);?/g, 'require($1);')
    .replace(/const \* as /g, 'const ')
    .replace(
      /export \{(.*?)\}\s*from\s*(['"].*?['"]);?/gs,
      (a, exports, from) => {
        const importAs = exports.replace(/\s*as\s*/g, `:${s}`);
        const imp = `const${s}{${importAs}}${s}=${s}require(${from});`
        const matches = exports.split(',')
          .map((s) => s.trim())
          .map((m) => m.replace(/(?:\w+\s+as\s+)(.*)/, '$1'))
          .map((m) => `module.exports.${m}${s}=${s}${m};`);
        return `${imp}${s}${matches.join(s)}`;
      }
    )
    .replace(/export\s+default\s+/g, `module.exports${s}=${s}`)
    .replace(/export\s+const\s+(\w+)/g, `const $1${s}=${s}module.exports.$1`)
    .replace(
      /export function (\w+)/g,
      `module.exports.$1${s}=${s}$1;function $1`
    )
    .replace(/\{(\w+) as (\w+)\}/g, `{$1:${s}$2}`)
  ;

  const searchPaths = [];
  output.code.replace(/require\(["'](.*)['"]\)/g, (s, match) => searchPaths.push(match));
  const searchRegex = new RegExp(`(?:${searchPaths.join('|')})`);

  paths
    .filter((p) => !/.*\.spec\..*/.test(p))
    .filter((p) => searchRegex.test(p))
    .map((p) => [p, relative(path, p).replace('../', './').replace('./../', '../')])
    .map(([p, relative]) => [
      p.replace(`${srcDir}/`, '').replace(/\.[tj]s?$/g, ''),
      relative.endsWith('.ts') ? relative : `${relative ? `${relative}/` : './'}index.ts`
    ])
    .forEach(([absPath, localPath]) => {
      output.code = output.code.replace(`require("${absPath}")`, `require("${localPath}.js")`);
    });
  return output;
}

async function transformBabel(source, path) {
  return babel.transformAsync(source, {
    filename: path,
    minified: process.env.NODE_ENV === 'production',
    retainLines: process.env.NODE_ENV !== 'production',
    ...babelConfig,
  });
}

async function build() {
  const start = Date.now();
  const paths = await glob(`${srcDir}/**`);
  // const srcPaths = await glob(`${srcDir}/utils/hash.ts`);
  await rmdir(distDir, { recursive: true });
  await Promise.all(
    paths
      .filter((path) => !/\.(test|spec|d)\.ts/.test(path))
      .filter((path) => /\.(ts|json)$/.test(path))
      .map(async (path) => ({ path, file: await readFile(path) }))
      .map(async (res) => {
        const { path, file } = await res;
        const fileText = file.toString();
        const newPath = path.replace(rootDir, '<rootDir>')
          .replace('src', 'dist')
          .replace('<rootDir>', rootDir)
          .replace(/\.ts$/, '.js');
        if (!existsSync(dirname(newPath))) {
          await mkdir(dirname(newPath), { recursive: true });
        }
        if (/\.json$/.test(path)) {
          await writeFile(newPath, fileText);
          return;
        }
        let result = fileText;
        try {
          result = await transformEsbuild(fileText, path, paths)
        } catch (e) {
          result = await transformBabel(fileText, path);
        }
        if (!result || !result.code) {
          console.error(`${path} failed to be transpiled.`);
          return;
        }
        if (['\n', '"use strict";\n', ''].includes(result.code)) {
          return;
        }
        const code = result.code.replace(/\.ts/g, '');
        await writeFile(newPath, code);
      }),
  );
  const timeTaken = (Date.now() - start) / 1000;
  console.log(` ✨ Took ${timeTaken}s to build. ✨`);
}

build().then(() => process.exit(0)).catch((e) => {
  console.error(e);
  process.exit(1);
});
