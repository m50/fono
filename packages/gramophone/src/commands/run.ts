import chalk from 'chalk';
import setupServer from 'setup/server';
import up from 'commands/migrate/up';
import { housekeeping } from 'housekeeping';

const PORT = process.env.PORT ?? 3000;
const URL = process.env.BASE_URL ?? `http://127.0.0.1:${PORT}`;

export const command = 'start:gramophone';
export const describe = 'Runs the backend `gramophone` server.';
export const builder = {};

export const handler = async () => {
  const app = setupServer();
  try {
    await up();
    housekeeping();
    console.log(`\n\tðŸ”Š Server started at ${chalk.cyan(`${URL}/`)} ðŸ”Š\n`);
    await app.listen(PORT, '0.0.0.0');
  } catch (err) {
    app.log.error(err);
    process.exit(1);
  }
};
