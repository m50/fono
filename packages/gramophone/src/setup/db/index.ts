import { knex, Knex } from 'knex';
import castTimestamps from './castTimestamps';
import castJson from './castJson';
import getRelationships from './relations';

type Result = Record<string, any> | Record<string, any>[] | undefined | null;

const config: Record<string, Knex.Config> = {
  test: {
    client: 'sqlite3',
    connection: {
      filename: ':memory:',
    },
    useNullAsDefault: true,
  },
  production: {
    client: 'mysql',
    connection: {
      host: process.env.DB_HOST,
      user: process.env.DB_USER,
      password: process.env.DB_PASSWORD,
      database: process.env.DB_DATABASE,
    },
  },
};

const db = knex({
  ...(config[process.env.NODE_ENV ?? 'test'] ?? config.production),
  postProcessResponse(result: Result, context: Record<string, any>) {
    if (!result || (typeof result !== 'object' && !Array.isArray(result))) {
      return result;
    }

    if (Array.isArray(result)) {
      return result.map((record) => getRelationships(castTimestamps(castJson(record)), context));
    }

    return getRelationships(castTimestamps(castJson(result)), context);
  },
});

knex.QueryBuilder.extend('withRelations', function withRelations(...relations: string[]) {
  // eslint-disable-next-line
  const { table } = this._single;
  return this.queryContext({ relations, db, table });
});

knex.QueryBuilder.extend('maybeWhere', function maybeWhere(column, value) {
  if (typeof value === 'undefined') {
    return this;
  }

  return this.where(column, value);
});

export default db;
