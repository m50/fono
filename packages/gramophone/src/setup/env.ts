import dotenv from 'dotenv';
import dotenvExpand from 'dotenv-expand';
import { join } from 'path';
import { existsSync as exists } from 'fs';

const files = [
  '.env',
  '.env.local',
];

files.forEach((file) => {
  if (exists(join(__dirname, '..', '..', file))) {
    const env = dotenv.config({ path: join(__dirname, '..', '..', file) });
    dotenvExpand(env);
  }
  if (exists(join(__dirname, '..', '..', '..', file))) {
    const env = dotenv.config({ path: join(__dirname, '..', '..', '..', file) });
    dotenvExpand(env);
  }
});
