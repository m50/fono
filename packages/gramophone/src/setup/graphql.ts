import { ApolloServer, IResolvers } from 'apollo-server-fastify';
import { glob } from 'glob';
import { basename, dirname } from 'path';
import type { DocumentNode } from 'graphql';
import type { FastifyInstance } from 'fastify';

export default (app: FastifyInstance<any, any, any>) => {
  const typeDefs = glob.sync(`${dirname(__dirname)}/graphql/schema/**`, { dot: true, nodir: true })
    .map((p) => require(p))
    .map((node: DocumentNode) => node);

  const resolvers: IResolvers = {}
  glob.sync(`${dirname(__dirname)}/graphql/resolvers/**`, { dot: true, nodir: true })
    .map((p) => [p, require(p)])
    .forEach(([p, resolver]) => {
      const resolverName = basename(p).replace(/\.\w+$/, '');
      resolvers[resolverName] = resolver;
    })

  const apollo = new ApolloServer({
    typeDefs,
    resolvers,
  });
  app.register(apollo.createHandler());

  apollo.setGraphQLPath('/ql'); // /g/ql

  return app;
};
