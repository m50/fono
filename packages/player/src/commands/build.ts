import { Arguments, Argv } from 'yargs';
import { join } from 'path';
import { build as execute } from '../build/build';

interface Args {
  rootDir: string;
  debug: boolean;
  prod: boolean;
}

export const command = 'build [rootDir]';
export const describe = 'Builds a directory using a set of configured transformers.';
export const builder = (yargs: Argv): Argv<Args> => {
  return yargs.positional('rootDir', {
    type: 'string',
    default: './',
    description: 'The root directory of the project to build. Default is the current working directory.',
    demandOption: false,
  }).option('debug', {
    type: 'boolean',
    alias: [
      'verbose',
      'v'
    ],
    default: false,
    description: 'Output additional debug information into build log.',
    demandOption: false,
  }).option('prod', {
    type: 'boolean',
    alias: 'p',
    default: false,
    description: 'Create a production build (minimized).',
    demandOption: false,
  });
};

export const handler = async (argv: Arguments<Args>) => {
  if (argv.prod) {
    process.env.NODE_ENV = 'production';
  }
  await execute(join(process.cwd(), argv.rootDir).replace(/\/$/, ''), argv.debug);
}
