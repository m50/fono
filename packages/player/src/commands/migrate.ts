import { reset, rollback, up, create } from '@fono/gramophone/src/commands/migrate';
import type { Argv } from 'yargs';

export default function setup(yargs: Argv) {
  return yargs.command('migrate:up', 'Runs any migrations that haven\'t been run yet.', (yargs) => {
    return yargs.option('quiet', {
      type: 'boolean',
      describe: 'Silently perform the migration.',
      alias: 'q',
    });
  }, async ({ quiet }) => await up(quiet))
    .command(['migrate:down', 'migrate:rollback'], 'Rolls back the newest set of migrations.', (yargs) => {
      return yargs.option('quiet', {
        type: 'boolean',
        describe: 'Silently perform the migration.',
        alias: 'q',
      });
    }, async ({ quiet }) => await rollback(quiet))
    .command('migrate:reset', 'Completely resets the database, undoing all migrations.', (yargs) => {
      return yargs.option('quiet', {
        type: 'boolean',
        describe: 'Silently perform the migration.',
        alias: 'q',
      })
    }, async ({ quiet }) => await reset(quiet))
    .command('migrate:create <tableName>', 'Creates a new migration.', (yargs) => {
      return yargs.option('quiet', {
        type: 'boolean',
        describe: 'Silently perform the migration.',
        alias: 'q',
      }).positional('tableName', {
        type: 'string',
        describe: 'The name of the table to create.',
        required: true,
      });
    }, async ({ quiet, tableName }) => await create(tableName as string, quiet));
}
