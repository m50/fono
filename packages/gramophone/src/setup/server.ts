import fastify from 'fastify';
import cors from 'fastify-cors';
import helmet from 'fastify-helmet';
import { setup as sockets } from 'sockets';
import * as endpoints from 'endpoints';
import { register as graphql } from './graphql';

export default () => {
  const server = fastify({ logger: true });
  server.register(cors);
  server.register(helmet);
  server.register(graphql);

  // Health check endpoint.
  server.register(async (app, _, done) => {
    app.get('/ping', async (request, reply) => {
      reply.type('application/json').code(200);
      return { message: 'pong' };
    });

    done();
  });

  // Setup Web Sockets.
  server.register(sockets);

  // Setup all over endpoints.
  Object.values(endpoints)
    .forEach((endpoint) => server.register(endpoint));

  return server;
};
