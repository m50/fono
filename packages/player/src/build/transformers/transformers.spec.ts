import { dirname, join } from 'path';
import glob from 'glob-promise';
import { rmSync as rm } from 'fs';
import { transform as regex } from './es5-regex';
import { transform as paths } from './paths';
import { transform as esbuild } from './esbuild';
import { transform as babel } from './babel';
import { transform as stringify } from './stringify';
import { transform as graphql } from './graphql';
import { transform } from '.';

const rootDir = dirname(dirname(dirname(__dirname)));
const input = `
import { transform as regex } from './es5-regex';
import { command } from 'commands/build';
import * as Test from '@fono/gramophone/src/setup/graphql';
import React from 'react';
export { k as test } from './test';

export const k: string = \`hi\`;
export default (k: number): void => {

};
export async function blah<P>(l: P) {}
export function blah2() {}
`;

describe('transformers', () => {
  beforeEach(() => { process.env.NODE_ENV = 'test'; });
  afterAll(() => rm(join(dirname(dirname(dirname(__dirname))), 'dist', 'discovery.json')));

  describe('es5-regex', () => {
    it('matches snapshot', async () => {
      const output = await regex(input, __filename);
      expect(output).toMatchSnapshot();
    });
    it('matches snapshot PRODUCTION', async () => {
      process.env.NODE_ENV = 'production';
      const output = await regex(input, __filename);
      expect(output).toMatchSnapshot();
    });
  });

  describe('paths', () => {
    it('matches snapshot', async () => {
      const p = await glob(`${rootDir}/src/**`, { nodir: true });
      const result = await paths(await regex(input, __filename), __filename, rootDir, p);
      expect(/\.\.\/commands\/build.js/.test(result)).toBeTruthy();
      expect(/'@fono\/gramophone\/dist\/setup\/graphql'/.test(result)).toBeTruthy();
      expect(result).toMatchSnapshot();
    });
  });

  describe('stringify', () => {
    it('matches snapshot', async () => {
      const output = await stringify(input, __filename);
      expect(output).toMatchSnapshot();
    });
  });

  describe('esbuild', () => {
    it('matches snapshot', async () => {
      const output = await esbuild(input, __filename, '{}');
      expect(output).toMatchSnapshot();
    });
    it('matches snapshot PRODUCTION', async () => {
      process.env.NODE_ENV = 'production';
      const output = await esbuild(input, __filename, '{}');
      expect(output).toMatchSnapshot();
    });
  });

  describe('babel', () => {
    it('matches snapshot', async () => {
      const output = await babel(input, __filename, rootDir);
      expect(output).toMatchSnapshot();
    });
    it('matches snapshot PRODUCTION', async () => {
      process.env.NODE_ENV = 'production';
      const output = await babel(input, __filename, rootDir);
      expect(output).toMatchSnapshot();
    });
  });

  describe('graphql', () => {
    it('matches snapshot', async () => {
      const inp = `
type Query {
  user: User
}
`;
      const output = await graphql(inp, __filename);
      expect(output).toMatchSnapshot();
    });
  });

  describe('transform root function', () => {
    it('gets a function even with invalid type', async () => {
      const p = await glob(`${rootDir}/src/**`, { nodir: true });
      const tr = transform('not-a-type', rootDir, p);
      expect(await tr(input, __filename)).toBe(input);
    });
  });
});

