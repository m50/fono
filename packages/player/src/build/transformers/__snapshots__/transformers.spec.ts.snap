// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformers babel matches snapshot 1`] = `
"\\"use strict\\";Object.defineProperty(exports, \\"__esModule\\", { value: true });exports.blah = blah;exports.blah2 = blah2;Object.defineProperty(exports, \\"test\\", { enumerable: true, get: function () {return _test.k;} });exports.default = exports.k = void 0;




var _test = require(\\"./test\\");

const k = \`hi\`;exports.k = k;var _default =
k => {

};exports.default = _default;
async function blah(l) {}
function blah2() {}"
`;

exports[`transformers babel matches snapshot PRODUCTION 1`] = `"\\"use strict\\";Object.defineProperty(exports,\\"__esModule\\",{value:true});exports.blah=blah;exports.blah2=blah2;Object.defineProperty(exports,\\"test\\",{enumerable:true,get:function(){return _test.k}});exports.default=exports.k=void 0;var _test=require(\\"./test\\");const k=\`hi\`;exports.k=k;var _default=k=>{};exports.default=_default;async function blah(l){}function blah2(){}"`;

exports[`transformers es5-regex matches snapshot 1`] = `
"'use strict';
const { transform: regex } = require('./es5-regex');
const { command } = require('commands/build');
const Test = require('@fono/gramophone/src/setup/graphql');
const React = require('react');
const { k: test } = require('./test'); module.exports.test = test;

const k = module.exports.k: string = \`hi\`;
module.exports.default = module.exports = (k: number): void => {

};
module.exports.blah = blah;async function blah<P>(l: P) {}
module.exports.blah2 = blah2;function blah2() {}
"
`;

exports[`transformers es5-regex matches snapshot PRODUCTION 1`] = `
"'use strict';
const { transform:regex }=require('./es5-regex');
const { command }=require('commands/build');
const Test=require('@fono/gramophone/src/setup/graphql');
const React=require('react');
const{ k:test }=require('./test');module.exports.test=test;

const k=module.exports.k: string = \`hi\`;
module.exports.default=module.exports=(k: number): void => {

};
module.exports.blah=blah;async function blah<P>(l: P) {}
module.exports.blah2=blah2;function blah2() {}
"
`;

exports[`transformers esbuild matches snapshot 1`] = `
"export {k as test} from \\"./test\\";
export const k = \`hi\`;
export default (k3) => {
};
export async function blah(l) {
}
export function blah2() {
}
"
`;

exports[`transformers esbuild matches snapshot PRODUCTION 1`] = `
"export{k as test}from\\"./test\\";export const k=\\"hi\\";export default o=>{};export async function blah(o){}export function blah2(){}
"
`;

exports[`transformers graphql matches snapshot 1`] = `
"'use strict';
// Original Definition at transformers/transformers.spec.ts
const { gql } = require('apollo-server-fastify');
module.exports = gql\`
type Query {
  user: User
}
\`;
module.exports.default = gql\`
type Query {
  user: User
}
\`;
module.exports.raw = \`
type Query {
  user: User
}
\`;
"
`;

exports[`transformers paths matches snapshot 1`] = `
"'use strict';
const { transform: regex } = require('./es5-regex');
const { command } = require('../../commands/build.js');
const Test = require('@fono/gramophone/dist/setup/graphql');
const React = require('react');
const { k: test } = require('./test'); module.exports.test = test;

const k = module.exports.k: string = \`hi\`;
module.exports.default = module.exports = (k: number): void => {

};
module.exports.blah = blah;async function blah<P>(l: P) {}
module.exports.blah2 = blah2;function blah2() {}
"
`;

exports[`transformers stringify matches snapshot 1`] = `
"module.exports = \`
import { transform as regex } from './es5-regex';
import { command } from 'commands/build';
import * as Test from '@fono/gramophone/src/setup/graphql';
import React from 'react';
export { k as test } from './test';

export const k: string = \`hi\`;
export default (k: number): void => {

};
export async function blah<P>(l: P) {}
export function blah2() {}
\`;
module.exports.default = \`
import { transform as regex } from './es5-regex';
import { command } from 'commands/build';
import * as Test from '@fono/gramophone/src/setup/graphql';
import React from 'react';
export { k as test } from './test';

export const k: string = \`hi\`;
export default (k: number): void => {

};
export async function blah<P>(l: P) {}
export function blah2() {}
\`;
"
`;
